src/tree/taffy_tree/tree.rs

```diff
@@ -49,6 +51,10 @@ pub struct Taffy {
     /// the layout algorithms during layout, while exposing the `NodeData.final_layout` when called by external users.
     /// This allows us to fix <https://github.com/DioxusLabs/taffy/issues/501> without breaking backwards compatibility
     pub(crate) is_layouting: bool,
+
+    /// Used to cache the resolved children of a node (taking into account `Display::contents`) during layout
+    /// so that repeated calls to the children method don't need to re-resolve the list.
+    node_children_cache: RefCell<(NodeId, Vec<NodeId>)>,
```

Comment:
Did you also try an approach that uses an enum which can switch between `Vec<NodeId>` and just `NodeId` for the non-content cases?
I'm not entirely sure, but I think that could avoid Heap allocation for the simple cases and maybe improve performance.
It might make the usage more complicated though, so I'm not sure if it's worth it.

Also, as we don't mutate the `Vec` after creating it, would `Box<[NodeId]>` also work?

Reply:
> Also, as we don't mutate the Vec after creating it, would Box<[NodeId]> also work? Then we save the capacity of the Vec which could slightly improve performance.

Ah, but we do mutate the Vec after creating it. We reuse the Vec each time we have to recompute children (usually because we're computing them for a different node). Which means we're actually doing very few allocations (only when a node's child count exceeds the max previously-seen child count)

> Did you also try an approach that uses an enum which can switch between Vec<NodeId> and just NodeId for the non-content cases?

The problem with this is that we don't know whether we're in a "simple" case (no `Display::Contents`) or a complex one until we've iterated through all the children and checked their display. If we don't copy all the `NodeId`s into a new Vec (well, the same re-used Vec as discussed) as we go, then we would have to backtrack and iterate over each child again in the complex cases. It *might* be faster if there aren't many `Display::Contents` nodes (which to be fair is likely), but I wouldn't like to bet on it.

Reply:
Ah I see, reusing the vec makes sense, my bad. Probably the best approach then